### ISSUE TITLE
Add ES6 Support

### ISSUE THREAD

**Issue by Constellation** [estools/escope#33]
Supporting ES6 block scopes and some special scopes (such as arraw function and method)

**Comment by Constellation** [comment]
__TODO__:
Need to support TDZ completely. Maybe instead of creating TDZ scope, marking reference as TDZ seems better.

**Comment by michaelficarra** [comment]
@Constellation: Yes, that is what we plan to do with shift-scope-js: https://github.com/shapesecurity/shift-scope-js/issues/17

**Comment by Constellation** [comment]
@michaelficarra:

Yeah, that's very cool. Agreed.

In the current version of `escope`, we explicitly create the `TDZ` scope for iteration-statements.
But creating TDZ scopes requires too many TDZ scopes (per let variable :P).
It's not suitable for users.
Tainting the reference as `TDZ` seems very nice solution :)

**Comment by nzakas** [comment]
@constellation almost there (only three failing tests!), but I might have found a bug. For this code:

```js
function bar() { q: for(;;) { break q; } }
```

I get a scope that has `q` in the `references` array:

```js
{ type: 'function',
  set: {},
  taints: {},
  dynamic: false,
  block:
   { type: 'FunctionDeclaration',
     id:
      { type: 'Identifier',
        name: 'bar',
        range: [Object],
        loc: [Object],
        parent: [Circular] },
     params: [],
     defaults: [],
     body:
      { type: 'BlockStatement',
        body: [Object],
        range: [Object],
        loc: [Object],
        parent: [Circular] },
     rest: null,
     generator: false,
     expression: false,
     range: [ 0, 42 ],
     loc: { start: [Object], end: [Object] },
     parent:
      { type: 'Program',
        body: [Object],
        range: [Object],
        loc: [Object],
        comments: [],
        tokens: [Object],
        parent: null } },
  through:
   [ { identifier: [Object],
       from: [Circular],
       tainted: false,
       resolved: null,
       flag: 1,
       __maybeImplicitGlobal: undefined } ],
  variables:
   [ { name: 'arguments',
       identifiers: [],
       references: [],
       defs: [],
       tainted: false,
       stack: true,
       scope: [Circular] } ],
  references:
   [ { identifier:    
         { type: 'Identifier',
           name: 'q',
           range: [ 17, 18 ],
           loc: { start: [Object], end: [Object] } },
       from: [Circular],
       tainted: false,
       resolved: null,
       flag: 1,
       __maybeImplicitGlobal: undefined } ],
// snip
```

So it looks like labels are incorrectly being added as in-scope identifiers. Can you confirm?

**Comment by nzakas** [comment]
Forgot to mention, I'm use 2.0.1.

**Comment by Constellation** [comment]
@nzakas 

Nice catch!! I'll fix it :)

**Comment by Constellation** [comment]
@nzakas 

OK, I've fixed the bug.
Which is the preferable for you?

1. update the `2.0.1` code with this fix.
2. publish the current code as `3.0.0`, currently `TDZ` is not updated yet. But other things are almost identical.

**Comment by nzakas** [comment]
If you can update 2.0.1 with the fix, I can keep testing for other incompatibilities.

**Comment by Constellation** [comment]
OK, so I've published it as `2.0.2`, it shipps the above bug fix :)

**Comment by nzakas** [comment]
Thanks!

**Comment by nzakas** [comment]
I found one more issue. It looks like escope isn't marking increment/decrement as write operations. For example:

```js
/*global b:false*/
b++;
```

The `ref` for this has `ref.isWrite()` return `false`, whereas in 1.x it returned `true`.

**Comment by Constellation** [comment]
@nzakas 
Cool! I'll fix it :)

**Comment by Constellation** [comment]
@nzakas 

I've fixed it and released it as `2.0.3` :)

**Comment by nzakas** [comment]
Thanks!

**Comment by nzakas** [comment]
@Constellation Found a couple more issues.

First, when using a label, references inside the label are not counted accurately:

```js
var foo = 5;

label: while (true) {
  console.log(foo);
  break label;
}
```

In this example, the variable `foo` incorrectly reports zero `references` when there should be one.

Second, block scope references don't appear to work in loops:

```js
for (let prop in box) {
    box[prop] = parseInt(box[prop]);
}
```

The variable `prop` also shows up with zero `references` instead of two.

**Comment by nzakas** [comment]
Sorry, I should have said `isRead()` was the problem, not `references`.

**Comment by nzakas** [comment]
Okay, nailed down the problem for the `let` in loops. Basically, the variable `prop` appears in the `variables` array for both the `TDZ` scope and the `block` scope. The reference to `prop` inside the `block` scope is then counted as a reference to `prop` in `block`, whereas the `prop` in `TDZ` is said to have zero references. This is what's confusing ESLint.

For the time being, I can avoid checking the TDZ for variables, but it seems like the TDZ logic is not quite correct.

**Comment by Constellation** [comment]
@nzakas 

Oh, thank you for your pointing.
These are raised when using `ecmaVersion: 6`?
I'll investigate this issues.

**Comment by nzakas** [comment]
Yes, with ecmaVersion: 6.

**Comment by jrvidal** [comment]
Just for reference, eslint/eslint#2405 is blocked by this issue, more concretely by "Initializers in Destructuring Assignments".
